File: kmp/log4agent/src/commonMain/kotlin/dev/log4agent/Log4AgentRedactor.kt
package dev.log4agent

object Log4AgentRedactor {
    const val REDACTED: String = "[REDACTED]"

    val defaultSensitiveKeys: Set<String> = setOf(
        "authorization",
        "auth",
        "bearer",
        "token",
        "access_token",
        "refresh_token",
        "password",
        "passwd",
        "secret",
        "cookie",
        "set-cookie",
        "session",
        "code",
    )

    fun redactAttributes(
        attributes: Map<String, String>,
        sensitiveKeys: Set<String> = defaultSensitiveKeys,
    ): Map<String, String> = attributes.mapValues { (key, value) ->
        if (isSensitiveKey(key, sensitiveKeys)) REDACTED else redactUrl(value, sensitiveKeys)
    }

    fun redactMessage(message: String, sensitiveKeys: Set<String> = defaultSensitiveKeys): String =
        redactUrl(message, sensitiveKeys)

    fun redactUrl(value: String, sensitiveKeys: Set<String> = defaultSensitiveKeys): String {
        if (!value.contains("?") && !value.contains("&")) return value
        var redacted = value
        sensitiveKeys.forEach { key ->
            val pattern = Regex("(?i)([?&]${Regex.escape(key)}=)[^&#\\s]*")
            redacted = redacted.replace(pattern) { match -> match.groupValues[1] + REDACTED }
        }
        return redacted
    }

    private fun isSensitiveKey(key: String, sensitiveKeys: Set<String>): Boolean {
        val normalized = key.lowercase()
        return sensitiveKeys.any { sensitive ->
            val candidate = sensitive.lowercase()
            normalized == candidate || normalized.contains(candidate)
        }
    }
}

File: flutter/lib/src/log4agent_redactor.dart
class Log4AgentRedactor {
  static const redacted = '[REDACTED]';

  static const defaultSensitiveKeys = <String>{
    'authorization',
    'auth',
    'bearer',
    'token',
    'access_token',
    'refresh_token',
    'password',
    'passwd',
    'secret',
    'cookie',
    'set-cookie',
    'session',
    'code',
  };

  static Map<String, String> redactAttributes(
    Map<String, String> attributes, {
    Set<String> keys = defaultSensitiveKeys,
  }) {
    return attributes.map((key, value) {
      if (_isSensitiveKey(key, keys)) return MapEntry(key, redacted);
      return MapEntry(key, redactUrl(value, keys: keys));
    });
  }

  static String redactUrl(
    String value, {
    Set<String> keys = defaultSensitiveKeys,
  }) {
    if (!value.contains('?') && !value.contains('&')) return value;
    var redactedValue = value;
    for (final key in keys) {
      final escaped = RegExp.escape(key);
      redactedValue = redactedValue.replaceAllMapped(
        RegExp('([?&]$escaped=)[^&#\\s]*', caseSensitive: false),
        (match) => '${match.group(1)}$redacted',
      );
    }
    return redactedValue;
  }

  static bool _isSensitiveKey(String key, Set<String> keys) {
    final normalized = key.toLowerCase();
    return keys.any((sensitive) {
      final candidate = sensitive.toLowerCase();
      return normalized == candidate || normalized.contains(candidate);
    });
  }
}

